pkg/apis/serving/v1/revision_helpers.go

```diff
@@ -87,6 +107,41 @@ func (rs *RevisionSpec) GetContainer() *corev1.Container {
 	return &corev1.Container{}
 }
 
+// SetRoutingState sets the routingState label on this Revision and updates the
+// routingStateModified annotation.
+func (r *Revision) SetRoutingState(state RoutingState) {
+	stateStr := string(state)
+	if t := r.ObjectMeta.Annotations[serving.RoutingStateModifiedAnnotationKey]; t != "" &&
+		r.Labels[serving.RoutingStateLabelKey] == stateStr {
+		return // Don't update timestamp if no change.
+	}
+
+	if r.Labels == nil {
+		r.Labels = make(map[string]string, 1)
```

Comment:
There might be a `UnionMaps` in `knative/pkg` somewhere - dunno if that makes things more succinct